### PR TITLE
Fix: Binary Maths

### PR DESCRIPTION
**Pull request by Tigh-Gherr**
Currently, arithmeitc that overflows `math.MaxInt64` and underflows `math.MinInt64` just causes integer overflows within the script, meaning maths that should be compatible with large numbers just does not work on our interpreter.

Also, `OP_NUM2BIN` and `OP_BIN2NUM` are implemented incorrectly.

To fix this, I've fixed the implementations of the two aforementioned opcodes, and changed the `scriptNum` data type to wrap ` big.Int`, implementing the correct math functions and byte conversion both to and from consensus, to and from standard big endian.

### CODE REVIEW

File: bscript/interpreter/number.go [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+// method to get the serialised representation (including values that overflow).
+//
+// Then, whenever data is interpreted as an integer, it is converted to this
+// type by using the NewNumber function which will return an error if the
+// number is out of range or not minimally encoded depending on parameters.
+// Since all numeric opcodes involve pulling data from the stack and
+// interpreting it as an integer, it provides the required behaviour.
+type scriptNumber struct {
+	val          *big.Int
+	afterGenesis bool
```
Comment (jadwahab) [inline bscript/interpreter/number.go:38]: is it only utxos which are after genesis that are treated this way?

File: bscript/interpreter/thread.go [hunk 2]
```diff
@@ -241,6 +241,7 @@ func (t *thread) CheckErrorCondition(finalScript bool) error {
 	if finalScript {
 		t.afterSuccess()
 	}
+
```
Comment (theflyingcodr) [inline bscript/interpreter/thread.go:244]: This is my fav change

File: bscript/interpreter/number.go (hunk 1 above)
Comment (Tigh-Gherr) [reply bscript/interpreter/number.go:38]: Nah every utxo, tho the ones before genesis are capped within the limits of pre-genesis

**Review by theflyingcodr — APPROVED** [review]
Yes

**Pull request merged**